File: webapp/src/songs.ts
// The premade song library. Every run picks one of these: randomly when
// offline, or whichever one the Archipelago seed chose (slot_data "song").
//
// The key and the backing track item names must match the apworld's
// Songs.py exactly.

export interface SongDefinition {
  key: string;   // stable id, shared with the apworld
  name: string;
  url: string;
  melodyTrack: number;   // the track the player reconstructs (always muted)
  backingTracks: { trackIndex: number; itemName: string }[];
  measureCount: number;
}

export const SONG_LIBRARY: SongDefinition[] = [
  {
    key: "korobeiniki",
    name: "Korobeiniki",
    url: "/songs/Korobeiniki.mid",
    melodyTrack: 0,
    backingTracks: [
      { trackIndex: 2, itemName: "Chord Track" },
      { trackIndex: 3, itemName: "Harmony Track" },
      { trackIndex: 5, itemName: "Bass Track" },
      { trackIndex: 6, itemName: "Drum Track" },
    ],
    measureCount: 56,
  },
  {
    key: "cancan",
    name: "Can-Can",
    url: "/songs/cancan.mid",
    melodyTrack: 0,
    backingTracks: [
      { trackIndex: 2, itemName: "Chord Track" },
      { trackIndex: 3, itemName: "Bass Track" },
      { trackIndex: 4, itemName: "Drum Track" },
    ],
    measureCount: 80,
  },
  {
    key: "flight_of_the_bumblebee",
    name: "Flight of the Bumblebee",
    url: "/songs/flight-of-the-bumblebee.mid",
    melodyTrack: 0,
    backingTracks: [
      { trackIndex: 2, itemName: "Chord Track" },
      { trackIndex: 4, itemName: "Bass Track" },
      { trackIndex: 5, itemName: "Drum Track" },
    ],
    measureCount: 101,
  },
  {
    key: "funiculi",
    name: "Funiculi Funicula",
    url: "/songs/funiculi.mid",
    melodyTrack: 0,
    backingTracks: [
      { trackIndex: 1, itemName: "Chord Track" },
      { trackIndex: 3, itemName: "Bass Track" },
      { trackIndex: 4, itemName: "Drum Track" },
    ],
    measureCount: 72,
  },
  {
    key: "william_tell",
    name: "William Tell Overture",
    url: "/songs/william-tell.mid",
    melodyTrack: 0,
    backingTracks: [
      { trackIndex: 1, itemName: "Chord Track" },
      { trackIndex: 2, itemName: "Bass Track" },
      { trackIndex: 3, itemName: "Drum Track" },
    ],
    measureCount: 80,
  },
  // add-song:entry (scripts/add-song.mjs inserts new songs here)
];

export function pickRandomSong(): SongDefinition {
  return SONG_LIBRARY[Math.floor(Math.random() * SONG_LIBRARY.length)];
}

export function songByKey(key: string): SongDefinition | undefined {
  return SONG_LIBRARY.find((song) => song.key === key);
}

File: composeapelago/Songs.py
# The premade song library. Each run randomly plays one of these songs,
# then tells the client through slot_data.
#
# The keys and track item names must match the webapp's src/songs.ts
# exactly. To add a song: compose the MIDI (see the webapp's
# scripts/make-starter-song.mjs), then run webapp/scripts/add-song.mjs.

# song key -> everything generation needs to know about it:
#   track_items   - which backing track items exist for it.
#   note_count    - melody notes; the location count in per_note mode.
#   bar_count     - melody bars; the location count in bars mode.
#   measure_count - measures shown/unlocked in the client.
SONGS = {
    "korobeiniki": {
        "name": "Korobeiniki",
        "track_items": ["Chord Track", "Harmony Track", "Bass Track", "Drum Track"],
        "note_count": 219,
        "bar_count": 56,
        "measure_count": 56,
    },
    "cancan": {
        "name": "Can-Can",
        "track_items": ["Chord Track", "Bass Track", "Drum Track"],
        "note_count": 426,
        "bar_count": 80,
        "measure_count": 80,
    },
    "flight_of_the_bumblebee": {
        "name": "Flight of the Bumblebee",
        "track_items": ["Chord Track", "Bass Track", "Drum Track"],
        "note_count": 683,
        "bar_count": 101,
        "measure_count": 101,
    },
    "funiculi": {
        "name": "Funiculi Funicula",
        "track_items": ["Chord Track", "Bass Track", "Drum Track"],
        "note_count": 324,
        "bar_count": 72,
        "measure_count": 72,
    },
    "william_tell": {
        "name": "William Tell Overture",
        "track_items": ["Chord Track", "Bass Track", "Drum Track"],
        "note_count": 813,
        "bar_count": 80,
        "measure_count": 80,
    },
    # add-song:song (scripts/add-song.mjs inserts new songs here)
}

